### PR TITLE
Added headers in scan http calls

### PR DESCRIPTION
**Pull request by shiv07tiwari**
## Description

Describe the changes made and why they were made instead of how they were made.

## Related issues and discussion
#fixes 2032
All header in all the http requests which use scan and barcode.
 
 ## Screen-shots, if any
 
 ## Checklist
 
Make sure you've done all the following (You can delete the checklist before submitting)
 
 - [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.
 - [x] Code is well documented
 - [x] Include unit tests for new functionality
 - [x] All user-visible strings are made translatable
 - [x] Code passes Travis builds in your branch
 - [x] If you have multiple commits please combine them into one commit by squashing them.
 - [x] Read and understood the contribution guidelines .

### CODE REVIEW

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIService.java [hunk 1]
```diff
@@ -21,14 +21,17 @@
 public interface OpenFoodAPIService {
 
     String PRODUCT_API_COMMENT = "Official Android app";
-
+    
+    @Headers("Open Food Facts Android App")
```
Comment (huzaifaiftikhar) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIService.java]: We can't use static headers like this. We need these to work for all the flavors of the app and not just OFF. Use dynamic header map instead with key "User-Agent" and value should be the app name with the version number.

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIService.java [hunk 2]
```diff
@@ -21,14 +21,17 @@
...
     @GET("api/v0/product/{barcode}.json?fields=image_small_url,vitamins_tags,minerals_tags,amino_acids_tags,other_nutritional_substances_tags,image_front_url,image_ingredients_url,image_nutrition_url,url,code,traces_tags,ingredients_that_may_be_from_palm_oil_tags,additives_tags,allergens_hierarchy,manufacturing_places,nutriments,ingredients_from_palm_oil_tags,brands_tags,traces,categories_tags,ingredients_text,product_name,generic_name,ingredients_from_or_that_may_be_from_palm_oil_n,serving_size,allergens_tags,allergens,origins,stores,nutrition_grade_fr,nutrient_levels,countries,countries_tags,brands,packaging,labels_tags,labels_hierarchy,cities_tags,quantity,ingredients_from_palm_oil_n,image_url,link,emb_codes_tags,states_tags,creator,created_t,last_modified_t,last_modified_by,editors_tags,nova_groups,lang,purchase_places,nutrition_data_per,no_nutrition_data")
     Call<State> getFullProductByBarcode(@Path("barcode") String barcode);
 
+    @Headers("Open Food Facts Android App")
```
Comment (huzaifaiftikhar) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIService.java]: Use dynamic header map

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIService.java [hunk 3]
```diff
@@ -21,14 +21,17 @@
...
     @GET("api/v0/product/{barcode}.json")
     Single<State> getExistingProductDetails(@Path("barcode") String barcode,
                                             @Query("fields") String fields);
 
+    @Headers("Open Food Facts Android App")
```
Comment (huzaifaiftikhar) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIService.java]: Use dynamic header map

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIService.java [hunk 4]
```diff
@@ -38,6 +41,7 @@
                                     @FieldMap Map<String, String> parameters,
                                     @Field("comment") String comment);
 
+    @Headers("Open Food Facts Android App")
```
Comment (huzaifaiftikhar) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIService.java]: Use dynamic header map

**Comment by huzaifaiftikhar** [comment]
@teolemon The scanner uses [this](https://github.com/openfoodfacts/openfoodfacts-androidapp/blob/master/app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIService.java#L32-L33) method. How to differentiate the headers? We can add "User-Agent" which can tell that API call has been made using the Official Android App. It could be a scan or a search as well.  @stephanegigandet What header can we add to distinguish different API calls from the same User-Agent? We can have the "User-Agent" during scan as "Official Android App xyz Scan" and during search we can change it to "Official Android App xyz Search", where xyz is the version number of the app. Is it a correct way of doing this?

**Comment by shiv07tiwari** [comment]
@huzaifaiftikhar  I understood the need to use dynamic headers.
Should I directly pass app name + version as a string ?

**Comment by teolemon** [comment]
@huzaifaiftikhar your solution looks good 👍

**Comment by shiv07tiwari** [comment]
@teolemon  Should I do this ?
1. Make a dynamic header in scanner method as "App Name + verison + Scan", and in other as "Name + version + Search" ?

**Comment by huzaifaiftikhar** [comment]
> Should I do this ?
> 
> 1. Make a dynamic header in scanner method as "App Name + verison + Scan", and in other as "Name + version + Search" ?

Sounds good to me 👍

**Comment by shiv07tiwari** [comment]
@huzaifaiftikhar  Should I add scan in getShortProductByBarcode() as well. I mean is this call also used during Scan, or search ?

**Comment by huzaifaiftikhar** [comment]
> @huzaifaiftikhar Should I add scan in getShortProductByBarcode() as well. I mean is this call also used during Scan, or search ?

No this is not used during scan. Refer [this](https://futurestud.io/tutorials/retrofit-2-dynamic-request-headers-with-headermap) for adding dynamic headers.

**Comment by shiv07tiwari** [comment]
@huzaifaiftikhar  Can't i set dynamic headers without map : 
// Replaces the header with the the value of its target.
@GET("/")
void foo(@Header("Accept-Language") String lang, Callback<Response> cb);
I mean just a string eg. "Official android app 2.3 Scan" in lang (Just an example) ?

**Comment by shiv07tiwari** [comment]
Also, to get Version, we need context, which I am not able to pass from test classes, but all these network classes are added in test, can you help me with it ?

@huzaifaiftikhar

**Comment by huzaifaiftikhar** [comment]
What if a REST method is used for both searching as well as scanning? Later on we can remove all the methods and just keep a single method for both scanning as well as searching. You might have already observed that both ```getFullProductByBarcode``` and ```getFullProductByBarcodeSingle``` are exactly the same API calls except that the former uses the Retrofit's default Call and the latter uses RxJava. So at some point of time we could just keep the Rx one and remove the other.

For getting the version name you can simply use the String ```BuildConfig.VERSION_NAME```

**Comment by shiv07tiwari** [comment]
@huzaifaiftikhar  I dont really understand how using the HeadersMap can solve the problem you mentioned, because i think the difference in my method and the blog is only about passing the parameters as a map instead of a string.
Sorry I am new to networking, so a bit confused about the stuff.

**Comment by huzaifaiftikhar** [comment]
Yes making the Header as a parameter of the method would fix our problem, no need for a map we can use a String as well since we have to add only a single header at present. Use "User-Agent" and not "Accept-Language"

**Comment by shiv07tiwari** [comment]
@huzaifaiftikhar  Made the changes.

**Comment by shiv07tiwari** [comment]
Sorry is closed it by mistake.

**Comment by teolemon** [comment]
```82.255.181.124 - - [31/Dec/2018:16:04:04 +0100] "GET /api/v0/product/5997523313111.json?fields=image_small_url,vitamins_tags,minerals_tags,amino_acids_tags,other_nutritional_substances_tags,image_front_url,image_ingredients_url,image_nutrition_url,url,code,traces_tags,ingredients_that_may_be_from_palm_oil_tags,additives_tags,allergens_hierarchy,manufacturing_places,nutriments,ingredients_from_palm_oil_tags,brands_tags,traces,categories_tags,ingredients_text,product_name,generic_name,ingredients_from_or_that_may_be_from_palm_oil_n,serving_size,allergens_tags,allergens,origins,stores,nutrition_grade_fr,nutrient_levels,countries,countries_tags,brands,packaging,labels_tags,labels_hierarchy,cities_tags,quantity,ingredients_from_palm_oil_n,image_url,link,emb_codes_tags,states_tags,creator,created_t,last_modified_t,last_modified_by,editors_tags,nova_groups,lang,purchase_places,nutrition_data_per,no_nutrition_data HTTP/2.0" 200 5742 "-" "Official Android App2.9.8Scan"```

**Comment by teolemon** [comment]
Spaces between the Name, Scan and Version Number

**Comment by teolemon** [comment]
``` 82.255.181.124 - - [31/Dec/2018:16:07:29 +0100] "GET /api/v0/product/5997523313111.json?fields=image_small_url,vitamins_tags,minerals_tags,amino_acids_tags,other_nutritional_substances_tags,image_front_url,image_ingredients_url,image_nutrition_url,url,code,traces_tags,ingredients_that_may_be_from_palm_oil_tags,additives_tags,allergens_hierarchy,manufacturing_places,nutriments,ingredients_from_palm_oil_tags,brands_tags,traces,categories_tags,ingredients_text,product_name,generic_name,ingredients_from_or_that_may_be_from_palm_oil_n,serving_size,allergens_tags,allergens,origins,stores,nutrition_grade_fr,nutrient_levels,countries,countries_tags,brands,packaging,labels_tags,labels_hierarchy,cities_tags,quantity,ingredients_from_palm_oil_n,image_url,link,emb_codes_tags,states_tags,creator,created_t,last_modified_t,last_modified_by,editors_tags,nova_groups,lang,purchase_places,nutrition_data_per,no_nutrition_data HTTP/2.0" 200 5742 "-" "Official Android App2.9.8Search" ```

**Comment by shiv07tiwari** [comment]
Fixed it @teolemon

**Comment by teolemon** [comment]
Normally that's what it now does. Cf the PR @huzaifaiftikhar

**Comment by teolemon** [comment]
```82.255.181.124 - - [31/Dec/2018:16:07:29 +0100] "GET /api/v0/product/5997523313111.json?fields=image_small_url,vitamins_tags,minerals_tags,amino_acids_tags,other_nutritional_substances_tags,image_front_url,image_ingredients_url,image_nutrition_url,url,code,traces_tags,ingredients_that_may_be_from_palm_oil_tags,additives_tags,allergens_hierarchy,manufacturing_places,nutriments,ingredients_from_palm_oil_tags,brands_tags,traces,categories_tags,ingredients_text,product_name,generic_name,ingredients_from_or_that_may_be_from_palm_oil_n,serving_size,allergens_tags,allergens,origins,stores,nutrition_grade_fr,nutrient_levels,countries,countries_tags,brands,packaging,labels_tags,labels_hierarchy,cities_tags,quantity,ingredients_from_palm_oil_n,image_url,link,emb_codes_tags,states_tags,creator,created_t,last_modified_t,last_modified_by,editors_tags,nova_groups,lang,purchase_places,nutrition_data_per,no_nutrition_data HTTP/2.0" 200 5742 "-" "Official Android App2.9.8Search"``` 

```
82.255.181.124 - - [31/Dec/2018:16:04:04 +0100] "GET /api/v0/product/5997523313111.json?fields=image_small_url,vitamins_tags,minerals_tags,amino_acids_tags,other_nutritional_substances_tags,image_front_url,image_ingredients_url,image_nutrition_url,url,code,traces_tags,ingredients_that_may_be_from_palm_oil_tags,additives_tags,allergens_hierarchy,manufacturing_places,nutriments,ingredients_from_palm_oil_tags,brands_tags,traces,categories_tags,ingredients_text,product_name,generic_name,ingredients_from_or_that_may_be_from_palm_oil_n,serving_size,allergens_tags,allergens,origins,stores,nutrition_grade_fr,nutrient_levels,countries,countries_tags,brands,packaging,labels_tags,labels_hierarchy,cities_tags,quantity,ingredients_from_palm_oil_n,image_url,link,emb_codes_tags,states_tags,creator,created_t,last_modified_t,last_modified_by,editors_tags,nova_groups,lang,purchase_places,nutrition_data_per,no_nutrition_data HTTP/2.0" 200 5742 "-" "Official Android App2.9.8Scan"
```

**Comment by shiv07tiwari** [comment]
@teolemon @huzaifaiftikhar  I dont understand. Is their any problem ?

**Comment by huzaifaiftikhar** [comment]
I was suggesting that for the case when it's neither a Scan nor a Search you have used the text "Unknown user agent". I am saying that we can change that text to "Official Android App x.y.z" without the Scan/Search. This way at least we are sure that the API was used by our official app.

**Comment by shiv07tiwari** [comment]
Oh okay, I'll do it @huzaifaiftikhar

**Comment by teolemon** [comment]
ah ok :-) 👍

**Comment by shiv07tiwari** [comment]
Fixed it @teolemon  @huzaifaiftikhar

**Comment by teolemon** [comment]
Can we also had the HEADER to the image calls as well as the taxonomies call ?

**Comment by huzaifaiftikhar** [comment]
Merged, thanks @shiv07tiwari

**Pull request merged**